File: Go/coverage/HashTableOpen.go
package datastructures

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"math"
	"os"
)

type HashEntry struct {
	key        int
	value      int
	isOccupied bool
	isDeleted  bool
}

type HashTableOpen struct {
	table    []HashEntry
	size     int
	capacity int
}

func NewHashTableOpen(initCap int) *HashTableOpen {
	if initCap <= 0 {
		initCap = 8
	}
	table := make([]HashEntry, initCap)
	for i := 0; i < initCap; i++ {
		table[i] = HashEntry{isOccupied: false, isDeleted: false}
	}
	return &HashTableOpen{
		table:    table,
		size:     0,
		capacity: initCap,
	}
}

func (h *HashTableOpen) hash(key int) int {
	return int(math.Abs(float64(key))) % h.capacity
}

func (h *HashTableOpen) resize() {
	oldCapacity := h.capacity
	oldTable := h.table

	h.capacity *= 2
	h.table = make([]HashEntry, h.capacity)
	for i := 0; i < h.capacity; i++ {
		h.table[i] = HashEntry{isOccupied: false, isDeleted: false}
	}

	h.size = 0
	for i := 0; i < oldCapacity; i++ {
		if oldTable[i].isOccupied && !oldTable[i].isDeleted {
			h.Insert(oldTable[i].key, oldTable[i].value)
		}
	}
}

func (h *HashTableOpen) Insert(key, value int) {
	if float64(h.size) >= float64(h.capacity)*0.7 {
		h.resize()
	}

	idx := h.hash(key)
	startIdx := idx

	for h.table[idx].isOccupied && !h.table[idx].isDeleted && h.table[idx].key != key {
		idx = (idx + 1) % h.capacity
		if idx == startIdx {
			return
		}
	}

	h.table[idx].key = key
	h.table[idx].value = value
	h.table[idx].isOccupied = true
	h.table[idx].isDeleted = false
	h.size++
}

func (h *HashTableOpen) Get(key int) (int, bool) {
	idx := h.hash(key)
	startIdx := idx

	for h.table[idx].isOccupied {
		if !h.table[idx].isDeleted && h.table[idx].key == key {
			return h.table[idx].value, true
		}
		idx = (idx + 1) % h.capacity
		if idx == startIdx {
			break
		}
	}
	return 0, false
}

func (h *HashTableOpen) Remove(key int) {
	idx := h.hash(key)
	startIdx := idx

	for h.table[idx].isOccupied {
		if !h.table[idx].isDeleted && h.table[idx].key == key {
			h.table[idx].isDeleted = true
			h.size--
			return
		}
		idx = (idx + 1) % h.capacity
		if idx == startIdx {
			break
		}
	}
}

func (h *HashTableOpen) Print() {
	fmt.Println("HashTableOpen:")
	for i := 0; i < h.capacity; i++ {
		if h.table[i].isOccupied && !h.table[i].isDeleted {
			fmt.Printf("[%d]: %d -> %d\n", i, h.table[i].key, h.table[i].value)
		}
	}
}

// Binary Serialization
func (h *HashTableOpen) Serialize(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for writing: %w", err)
	}
	defer file.Close()

	if err := binary.Write(file, binary.LittleEndian, uint64(h.size)); err != nil {
		return err
	}
	if err := binary.Write(file, binary.LittleEndian, uint64(h.capacity)); err != nil {
		return err
	}

	for i := 0; i < h.capacity; i++ {
		if err := binary.Write(file, binary.LittleEndian, int32(h.table[i].key)); err != nil {
			return err
		}
		if err := binary.Write(file, binary.LittleEndian, int32(h.table[i].value)); err != nil {
			return err
		}
		if err := binary.Write(file, binary.LittleEndian, h.table[i].isOccupied); err != nil {
			return err
		}
		if err := binary.Write(file, binary.LittleEndian, h.table[i].isDeleted); err != nil {
			return err
		}
	}
	return nil
}

func (h *HashTableOpen) Deserialize(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for reading: %w", err)
	}
	defer file.Close()

	var size, capacity uint64
	if err := binary.Read(file, binary.LittleEndian, &size); err != nil {
		return err
	}
	if err := binary.Read(file, binary.LittleEndian, &capacity); err != nil {
		return err
	}

	h.size = int(size)
	h.capacity = int(capacity)
	h.table = make([]HashEntry, h.capacity)

	for i := 0; i < h.capacity; i++ {
		var key, value int32
		if err := binary.Read(file, binary.LittleEndian, &key); err != nil {
			return err
		}
		if err := binary.Read(file, binary.LittleEndian, &value); err != nil {
			return err
		}
		if err := binary.Read(file, binary.LittleEndian, &h.table[i].isOccupied); err != nil {
			return err
		}
		if err := binary.Read(file, binary.LittleEndian, &h.table[i].isDeleted); err != nil {
			return err
		}
		h.table[i].key = int(key)
		h.table[i].value = int(value)
	}
	return nil
}

// JSON Serialization
type openEntry struct {
	Key   int `json:"key"`
	Value int `json:"value"`
}

type hashTableOpenJSON struct {
	Entries []openEntry `json:"entries"`
}

func (h *HashTableOpen) SerializeJSON(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for writing: %w", err)
	}
	defer file.Close()

	entries := make([]openEntry, 0, h.size)
	for i := 0; i < h.capacity; i++ {
		if h.table[i].isOccupied && !h.table[i].isDeleted {
			entries = append(entries, openEntry{Key: h.table[i].key, Value: h.table[i].value})
		}
	}

	data := hashTableOpenJSON{Entries: entries}
	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "  ")
	return encoder.Encode(data)
}

func (h *HashTableOpen) DeserializeJSON(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return fmt.Errorf("cannot open file for reading: %w", err)
	}
	defer file.Close()

	// Clear existing data
	h.table = make([]HashEntry, h.capacity)
	for i := 0; i < h.capacity; i++ {
		h.table[i] = HashEntry{isOccupied: false, isDeleted: false}
	}
	h.size = 0

	var data hashTableOpenJSON
	decoder := json.NewDecoder(file)
	if err := decoder.Decode(&data); err != nil {
		return err
	}

	for _, entry := range data.Entries {
		h.Insert(entry.Key, entry.Value)
	}
	return nil
}

File: structures/HashTableOpen.cpp
#include "HashTableOpen.h"
#include <iostream>
#include <fstream>
#include <cmath>

// ==========================================
// HashTableOpen Implementation
// ==========================================

HashTableOpen::HashTableOpen(size_t init_cap) : size(0), capacity(init_cap) {
    table = new HashEntry[capacity];
    for (size_t i = 0; i < capacity; i++) {
        table[i].isOccupied = false;
        table[i].isDeleted = false;
    }
}

HashTableOpen::~HashTableOpen() {
    delete[] table;
}

size_t HashTableOpen::hash(int key) const {
    return std::abs(key) % capacity;
}

void HashTableOpen::resize() {
    size_t old_capacity = capacity;
    HashEntry* old_table = table;
    
    capacity *= 2;
    table = new HashEntry[capacity];
    for (size_t i = 0; i < capacity; i++) {
        table[i].isOccupied = false;
        table[i].isDeleted = false;
    }
    
    size = 0;
    for (size_t i = 0; i < old_capacity; i++) {
        if (old_table[i].isOccupied && !old_table[i].isDeleted) {
            insert(old_table[i].key, old_table[i].value);
        }
    }
    
    delete[] old_table;
}

void HashTableOpen::insert(int key, int value) {
    if (size >= capacity * 0.7) {
        resize();
    }
    
    size_t idx = hash(key);
    size_t start_idx = idx;
    
    while (table[idx].isOccupied && !table[idx].isDeleted && table[idx].key != key) {
        idx = (idx + 1) % capacity;
        if (idx == start_idx) return;
    }
    
    table[idx].key = key;
    table[idx].value = value;
    table[idx].isOccupied = true;
    table[idx].isDeleted = false;
    size++;
}

std::optional<int> HashTableOpen::get(int key) {
    size_t idx = hash(key);
    size_t start_idx = idx;
    
    while (table[idx].isOccupied) {
        if (!table[idx].isDeleted && table[idx].key == key) {
            return table[idx].value;
        }
        idx = (idx + 1) % capacity;
        if (idx == start_idx) break;
    }
    
    return std::nullopt;
}

void HashTableOpen::remove(int key) {
    size_t idx = hash(key);
    size_t start_idx = idx;
    
    while (table[idx].isOccupied) {
        if (!table[idx].isDeleted && table[idx].key == key) {
            table[idx].isDeleted = true;
            size--;
            return;
        }
        idx = (idx + 1) % capacity;
        if (idx == start_idx) break;
    }
}

void HashTableOpen::print() const {
    std::cout << "HashTableOpen:\n";
    for (size_t i = 0; i < capacity; i++) {
        if (table[i].isOccupied && !table[i].isDeleted) {
            std::cout << "[" << i << "]: " << table[i].key << " -> " << table[i].value << "\n";
        }
    }
}

void HashTableOpen::serialize(const std::string& filename) const {
    std::ofstream file(filename, std::ios::binary);
    if (!file) throw std::runtime_error("Cannot open file for writing");
    
    file.write(reinterpret_cast<const char*>(&size), sizeof(size));
    file.write(reinterpret_cast<const char*>(&capacity), sizeof(capacity));
    
    for (size_t i = 0; i < capacity; i++) {
        file.write(reinterpret_cast<const char*>(&table[i].key), sizeof(int));
        file.write(reinterpret_cast<const char*>(&table[i].value), sizeof(int));
        file.write(reinterpret_cast<const char*>(&table[i].isOccupied), sizeof(bool));
        file.write(reinterpret_cast<const char*>(&table[i].isDeleted), sizeof(bool));
    }
    
    file.close();
}

void HashTableOpen::deserialize(const std::string& filename) {
    std::ifstream file(filename, std::ios::binary);
    if (!file) throw std::runtime_error("Cannot open file for reading");
    
    file.read(reinterpret_cast<char*>(&size), sizeof(size));
    file.read(reinterpret_cast<char*>(&capacity), sizeof(capacity));
    
    delete[] table;
    table = new HashEntry[capacity];
    
    for (size_t i = 0; i < capacity; i++) {
        file.read(reinterpret_cast<char*>(&table[i].key), sizeof(int));
        file.read(reinterpret_cast<char*>(&table[i].value), sizeof(int));
        file.read(reinterpret_cast<char*>(&table[i].isOccupied), sizeof(bool));
        file.read(reinterpret_cast<char*>(&table[i].isDeleted), sizeof(bool));
    }
    
    file.close();
}
